atd-toolbox/s3_restore_cr3_pdfs/s3_restore_cr3_pdfs.py

```diff
@@ -0,0 +1,268 @@
+from __future__ import print_function
+
+import re
+import os
+import sys
+import json
+import boto3
+import magic
+import argparse
+import datetime
+import requests
+from operator import attrgetter
+from botocore.config import Config
+
+# https://stackoverflow.com/questions/5574702/how-to-print-to-stderr-in-python
+def eprint(*args, **kwargs):
+    print(*args, file=sys.stderr, **kwargs)
+
+s3_resource = boto3.resource('s3')
+
+
+ACCESS_KEY = os.getenv('AWS_ACCESS_KEY_ID')
+SECRET_KEY = os.getenv('AWS_SECRET_ACCESS_KEY')
+HASURA_ADMIN_KEY = os.getenv('HASURA_ADMIN_KEY')
+HASURA_ENDPOINT = os.getenv('HASURA_ENDPOINT')
+
+HEADERS = {
+          "X-Hasura-Admin-Secret": HASURA_ADMIN_KEY,
+          "Content-Type": "application/json",
+          }
+
+bucket = 'atd-vision-zero-editor'
+
+
+# https://github.com/cityofaustin/atd-airflow/blob/master/dags/python_scripts/atd_vzd_cr3_scan_pdf_records.py#L24
+def is_valid_metadata(metadata: dict) -> bool:
+    if metadata.get("last_update", None) is not None \
+       and metadata.get("file_size", None) is not None \
+       and metadata.get("mime_type", None) is not None \
+       and metadata.get("encoding", None) is not None:
+        return True
+    return False
+
+
+# setup and parse arguments
+try:
+    argparse = argparse.ArgumentParser(description = 'Utility to restore last valid PDF in S3 for ATD VZ')
+
+    argparse.add_argument("-p", "--production",
+            help = 'Specify the use of production environment',
+            action = 'store_true')
+
+    argparse.add_argument("-c", "--crashes",
+            help = 'Specify JSON file containing crashes to operate on. Format: { "crashes": [ crash_id_0, crash_id_1, .. ] }',
+            required=True,
+            metavar = 'crashes.json')
+
+    argparse.add_argument("--i-understand",
+            help = 'Do not ask the user to acknoledge that this program changes the state of S3 objects and the database.',
+            action = 'store_true')
+
+    args = argparse.parse_args()
+except:
+    # a stderr log is not needed, argparse croaks verbosly
+    sys.exit(1)
```

Comment:
@frankhereford @sergiogcx i assume the idea is that Airflow will pickup the exit code (`1`) and handle alerts as needed? I've never actually attempted this in a docker container but it makes perfect sense :)

Reply:
Airflow should pick it up, most processes that have an exception or an error exit with a code greater than 0. By default Airflow will run the DAG through (as it finishes the execution), and depending on how it's written it will trigger the `on_failure_callback `, which in our case it drops a message in Slack.

Reply:
I wasn't actually think of this program ever running in Airflow. If it needs to, I'd like to move it into the ETL directory in the repo and test it out in the context of docker.

As it stands, it's a one-off script to fix a known set of crashes that have had their CR3 file over-written with an HTML page on the S3 bucket where they are stored. FWIW, here are [some JSON](https://gist.github.com/frankhereford/19017a97a628d2204a9a74adcb2afeaa) files - one for testing and one that defines that set of crashes.

Reply:
thank you both. i wasn't sure of the deployment/use context!